components/script/dom/shadowroot.rs

```diff
@@ -78,13 +79,25 @@ impl ShadowRoot {
     }
 
     pub fn detach(&self) {
-        self.host.set(None);
+        self.document.unregister_shadow_root(&self);
         let node = self.upcast::<Node>();
         node.set_containing_shadow_root(None);
-        node.set_flag(NodeFlags::IS_CONNECTED, false);
         for child in node.traverse_preorder(ShadowIncluding::No) {
-            child.set_flag(NodeFlags::IS_CONNECTED, false);
+            if node.RemoveChild(&child).is_err() {
```

Comment:
You don't really need `traverse_preorder` right? Just `GetFirstChild()` / `GetNextSibling()`.

Reply:
Also how well does `traverse_preorder` handle the DOM mutating while it's running? I wouldn't be surprised if "not too well" was the answer.